### PR TITLE
Improve GemLite Integration

### PR DESCRIPTION
**Pull request by mobicham**
### Tasks 
* Faster `get_plain()` via Triton unpacking instead of matmul with the identity matrix -> slicing should be faster. 
* Fix serialization issues.
* General clean-up.
* Remove cuda device check in `.to()` to allow loading models in transformers.
* Force cuda device in `from_plain` / `get_plain` for vllm weight loader compatibllity. 
* Add bfloat16 support.

Note: Slicing still performs unpacking -> packing. If we can restrict the step to be `self._layout.packing_bitwidth // self._layout.bit_width` we can avoid this and slice directly the packed data.
 
### Test
```Python
import torch, gemlite
from torchao.quantization import GemliteUIntXWeightOnlyConfig, quantize_
device = 'cuda:0'
dtype = torch.float16

layer = torch.nn.Linear(256, 512, bias=False, dtype=dtype, device=device)
weight = layer.weight.data.clone()
orig_shape = weight.shape

group_size = 64
quantize_(layer, GemliteUIntXWeightOnlyConfig(bit_width=4, group_size=group_size))

#Test dot prod
#################################################
torch.manual_seed(0)
x = torch.randn((1, layer.in_features), device=device, dtype=dtype) / 10.
y_ref = x @ weight.T
y_gem  = layer(x)
assert (y_ref - y_gem).abs().mean() < 5e-3, "Dot product mismatch"

#Test get_plain() 
#################################################
W_q, s, z       = layer.weight.tensor_impl.get_plain()
W_r_getplain = ((W_q.view([-1, group_size]) - z.view(-1, 1)) * s.view(-1, 1)).view(orig_shape)
W_r_dotprod = layer(torch.eye(layer.in_features, device=device, dtype=dtype)).T #SLOW because it autotunes
assert (W_r_getplain - W_r_dotprod).abs().mean() < 1e-4, "get_plain() incorrect results"
assert (y_gem - (x @ W_r_getplain.T)).abs().mean() < 1e-4, "get_plain() incorrect results"

#Test slicing 
#################################################
def _dequant(tensor_impl, in_features, orig_shape):
    int_data   = tensor_impl.packed_weight
    scale      = tensor_impl.scale
    zero_point = tensor_impl.zero_point

    W_q = gemlite.bitpack.unpack_over_rows(int_data, W_nbits=4, num_output_rows=in_features, dtype=torch.uint8).T.contiguous()
    s   = scale.t().contiguous()
    z   = zero_point.t().contiguous()
    return ((W_q.view([-1, group_size]) - z.view(-1, 1)) * s.view(-1, 1)).view(orig_shape)

torch.manual_seed(0)
x     = torch.randn((1, layer.in_features), device=device, dtype=dtype) / 10.
y_ref = layer(x)
W_r   = _dequant(layer.weight.tensor_impl, layer.in_features, orig_shape).T
y_2   = x @ W_r
assert (y_ref - y_2).abs().mean() < 1e-4, "Incorrect dequant results"


layer_sliced = layer.weight.narrow(0, 0, 256)
W_slice1   = _dequant(layer_sliced.tensor_impl, layer.in_features, [orig_shape[0]//2, orig_shape[1]]).T
assert (W_r[:, :256] - W_slice1).abs().mean() == 0, "slice1 along axis=0 is incorrect"

layer_sliced = layer.weight.narrow(0, 256, 256)
W_slice2   = _dequant(layer_sliced.tensor_impl, layer.in_features, [orig_shape[0]//2, orig_shape[1]]).T
assert (W_r[:, 256:] - W_slice2).abs().mean() == 0 , "slice2 along axis=0 is incorrect"

layer_sliced = layer.weight.narrow(1, 0, 128)
W_slice1   = _dequant(layer_sliced.tensor_impl, layer.in_features//2, [orig_shape[0], orig_shape[1]//2]).T
assert (W_r[:128, :] - W_slice1).abs().mean() == 0, "slice1 along axis=1 is incorrect"

layer_sliced = layer.weight.narrow(1, 128, 128)
W_slice2   = _dequant(layer_sliced.tensor_impl, layer.in_features//2, [orig_shape[0], orig_shape[1]//2]).T
assert (W_r[128:, :] - W_slice2).abs().mean() == 0 , "slice2 along axis=1 is incorrect"
```

### CODE REVIEW

File: test/dtypes/test_affine_quantized.py [hunk 1]
```diff
@@ -354,21 +355,18 @@ def test_slice(self, device, dtype):
         _ = dummy.weight.narrow(1, 0, 128)
 
     @common_utils.parametrize("device", ["cuda"])
-    @common_utils.parametrize("dtype", [torch.bfloat16])
```
Comment (jerryzh168) [inline test/dtypes/test_affine_quantized.py:357]: why is this test removed?

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 2]
```diff
@@ -172,48 +203,21 @@ def from_plain(
         zero_point: Optional[torch.Tensor],
         _layout: Layout,
     ):
-        from gemlite.core import DType, GemLiteLinearTriton, set_autotune
+        print(f"from plain: {int_data.shape=} {scale.shape=}")
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:206]: please remove print before landing

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 3]
```diff
@@ -316,37 +311,36 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
...
-                    zero_point = aten.slice.Tensor(
-                        zero_point, param_dim, start_scale, end_scale, step
-                    )
-                else:
-                    zero_point = None
-                sliced = self.from_plain(int_data, scale, zero_point, self._layout)
+                # TODO: maybe get_plain should not output a transposed scale and zp?
+                # since scale and zero_point are transposed from `get_plain`,
+                # we need to transpose them back before feeding to from_plain
+                scale = scale.t().contiguous()
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:321]: do we still need transpose?

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 4]
```diff
@@ -375,16 +371,21 @@ def _linear_fp_act_int4_weight_gemlite_impl(input_tensor, weight_tensor, bias=No
...
-
     batch_size = input_tensor.view(-1, input_tensor.shape[-1]).shape[0]
     matmul_type = _matmul_type_fn(batch_size, weight_impl._layout.bit_width)
 
+    if weight_impl.gemlite_kwargs["scaled_activations"]:
+        scale_activations = scale_activations_int8
+    else:
+        scale_activations = scale_activations_no_scaling
+
     return GemLiteLinearTriton.forward_functional(
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:382]: would this be another `getattr` that gets serialized?

File: test/dtypes/test_affine_quantized.py (hunk 1 above)
Comment (mobicham) [reply test/dtypes/test_affine_quantized.py:357]: Oh I have no idea, `ruff` removed it probably, I didn't manually touch that file

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 5]
```diff
@@ -246,34 +250,24 @@ def _apply_fn_to_data(self, fn):
...
     def get_plain(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
-        dq = (
-            _linear_fp_act_int4_weight_gemlite_impl(
-                torch.eye(
-                    self.scale.shape[0] * self._layout.group_size,
-                    device=self.device,
-                    dtype=self.scale.dtype,
-                ),
-                self,
+        elements_per_sample = self._layout.packing_bitwidth // self._layout.bit_width
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:253]: ideally I think get_plain() should return scale/zero_point in original shape, without the need to do transpose again to do from_plain:

```
int_data, scale, zp = tensor_impl.get_plain()
tensor_impl = TensorImpl.from_plain(int_data, scale, zp)
```

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 4 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:382]: This is what actually fixes the serialization issue. The activation scaling function was creating the issue when it was inside the state_dict. By moving it outside and putting it in `gemlite_layout.py` we no longer need to store the function for serialization.

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 5 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:253]: Sure, we just need to swap scale/zero axis in the slicing function then as referred to in https://github.com/pytorch/ao/pull/2096#discussion_r2054757000

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 2 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:206]: Will do

File: test/dtypes/test_affine_quantized.py (hunk 1 above)
Comment (mobicham) [reply test/dtypes/test_affine_quantized.py:357]: I am a bit confused where that `torch.bfloat16` comes from. I cloned from your repo and it was not there in the first place: https://github.com/pytorch/ao/blob/e23ec911f9a0a392cc0500b015e3a6879d5ba1ad/test/dtypes/test_affine_quantized.py#L357-L361

Normally, there shouldn't be `parametrize("dtype", [torch.bfloat16])` in the gemlite test  because it's hard-coded to work with float16 (gemlite support bfloat16 by the way).

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 5 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:252]: https://github.com/pytorch/ao/pull/2096/commits/6e52183a9939f9b08d1d09ece5b0d1d02183d571

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 2 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py]: https://github.com/pytorch/ao/commit/6e52183a9939f9b08d1d09ece5b0d1d02183d571

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 5 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:251]: https://github.com/pytorch/ao/pull/2096/commits/d5f03c6bb5645a05c0a75127c320a1ab02a49309

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 3 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py]: https://github.com/pytorch/ao/pull/2096/commits/d5f03c6bb5645a05c0a75127c320a1ab02a49309

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 6]
```diff
@@ -120,7 +152,7 @@ def __new__(
         _layout: Layout,
     ):
         kwargs = {}
-        kwargs["device"] = packed_weight.device
+        kwargs["device"] = "cuda"  # packed_weight.device
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:155]: can you restore this? maybe have an assert

**Review by jerryzh168 — APPROVED** [review]

File: torchao/dtypes/uintx/gemlite_layout.py (hunk 6 above)
Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py]: Oh my mistake, that was just for testing, reverting back to `packed_weight.device` 
https://github.com/pytorch/ao/pull/2096/commits/8fb8a26d2d2360552b0dcfa688612233c5c87fa0

File: torchao/dtypes/uintx/gemlite_layout.py [hunk 7]
```diff
@@ -246,34 +244,24 @@ def _apply_fn_to_data(self, fn):
...
+        elements_per_sample = self._layout.packing_bitwidth // self._layout.bit_width
+        in_features = (
+            self.packed_weight.numel() * elements_per_sample
+        ) // self.gemlite_kwargs["out_features"]
+        int_data = (
+            gemlite.bitpack.unpack_over_rows(
+                self.packed_weight,
+                W_nbits=self._layout.bit_width,
+                num_output_rows=in_features,
+                dtype=torch.uint8,
```
Comment (jerryzh168) [inline torchao/dtypes/uintx/gemlite_layout.py:256]: is transposed still needed here?

Comment (mobicham) [reply torchao/dtypes/uintx/gemlite_layout.py:256]: Yes because it's transposed during the packing process: `gemlite_linear(x) = x @ W_dequant.T`

File: test/dtypes/test_affine_quantized.py (hunk 1 above)
Comment (mobicham) [reply test/dtypes/test_affine_quantized.py:357]: It was in `test_slice_gemlite` actually, looks fine to me now:
```Python
    @common_utils.parametrize("device", ["cuda"])
    @common_utils.parametrize("dtype", [torch.bfloat16])
    @skip_if_no_cuda()
    def test_slice_int4wo(self, device, dtype):
        # in_feature not divisible by 1024
        # out_feature not divisible by 8
        # to test slice + padding for int4 weight only quantization
        dummy = nn.Linear(256, 321, dtype=dtype, device=device)
        quantize_(dummy, Int4WeightOnlyConfig())
        # make sure these run without error
        _ = dummy.weight.narrow(0, 0, 64)
        _ = dummy.weight.narrow(1, 0, 128)

    @common_utils.parametrize("device", ["cuda"])
    @common_utils.parametrize("dtype", [torch.float16])
    @skip_if_no_cuda()
    @skip_if_no_gemlite()
    def test_slice_gemlite(self, device, dtype):
        # in_feature not divisible by 1024
        # out_feature not divisible by 8
        # to test slice + padding for int4 weight only quantization
        dummy = nn.Linear(256, 512, dtype=dtype, device=device)
        quantize_(dummy, GemliteUIntXWeightOnlyConfig())
        # make sure these run without error
        _ = dummy.weight.narrow(0, 0, 64)
        _ = dummy.weight.narrow(1, 0, 128)
```

Comment (mobicham) [reply test/dtypes/test_affine_quantized.py:357]: For some reason `test_matmul` disappeared, I added it back.

File: torchao/testing/utils.py [hunk 8]
```diff
@@ -91,12 +91,30 @@ def wrapper(*args, **kwargs):
 
 
 def skip_if_no_cuda():
-    import unittest
+    import pytest
```
Comment (jerryzh168) [inline torchao/testing/utils.py:94]: we might have to use unittest to make CI happy I think
